invertedai/utils.py

```diff
@@ -563,6 +575,211 @@ def inside_fov(center: Point, initialize_fov: float, point: Point) -> bool:
     response.agent_attributes = agent_attributes
     return response
 
+def _get_drivable_area_ratio(centers,location,agent_density,scaling_factor):
+    try:
+        from cv2 import (
+            cvtColor,
+            COLOR_BGR2GRAY
+        )
+
+    except ImportError as e:
+        iai.logger.error(
+            "Calculating drivable area requires cv2, which was not found."
+        )
+        raise e
+
+    #Get fraction of image that is a drivable surface (assume all non-black is drivable)
+    center_road_area_dict = {}
+    max_drivable_area_ratio = 0
+    for area_center in tmap(
+        Point.fromlist, 
+        centers, 
+        total=len(centers),
+        desc=f"Calculating drivable surface areas"
+    ):
+        #Naively check every square within requested area
+        #TODO: Use heuristics or other methods to (e.g. map polygon, high FOV image, quadtree) to reduce computation time
+        center_tuple = (area_center.x, area_center.y)
+        birdview = iai.location_info(
+            location=location,
+            rendering_fov=100,
+            rendering_center=center_tuple
+        ).birdview_image.decode()
+
+        birdview = cvtColor(birdview,COLOR_BGR2GRAY)
+        number_of_black_pix = np.sum(birdview == 0)
+        h, w = birdview.shape
+        total_num_pixels = h*w
+
+        drivable_area_ratio = (total_num_pixels-number_of_black_pix)/total_num_pixels 
+        center_road_area_dict[center_tuple] = drivable_area_ratio
+
+        if drivable_area_ratio > max_drivable_area_ratio:
+            max_drivable_area_ratio = drivable_area_ratio
+
+    agent_density_list = []
+    for area_center, drivable_ratio in center_road_area_dict.items():
+        agent_density_list.append(_calculate_agent_density_max_scaled(agent_density,scaling_factor,drivable_ratio,max_drivable_area_ratio))
+
+    return agent_density_list
+
+def _calculate_agent_density_max_scaled(agent_density,scaling_factor,drivable_ratio,max_drivable_area_ratio):
+
+    return round(agent_density*(1-scaling_factor*(max_drivable_area_ratio-drivable_ratio)/max_drivable_area_ratio)) if drivable_ratio > 0.0 else 0
+
+def fill_initialization(
+    location, 
+    agent_density, 
+    traffic_lights_states=None, 
+    random_seed=None, 
+    map_center=(0, 0),
+    width=100, 
+    height=100, 
+    stride=50, 
+    initialize_fov=INITIALIZE_FOV, 
```

Comment:
It's weird to me that this is configurable, given that `initialize` itself does not allow fov to be changed.

Reply:
It looks like the "initialize_fov" parameter isn't actually passed into the initialize() in any way, it looks like it's used to determine which existing vehicles are to be considered as conditional for the current area. So to me, this is an issue of an unclear parameter name.